### PR TITLE
Js eager each

### PR DESCRIPTION
**Pull request by jonstorer**
A user can fetch more records than the limit imposed on a single request.

Continue to fetch until it has exhausted the results from the API or it has 300 records.
```ruby
response = Vindicia::Repository::Account.where({ limit: 300 })
```

Continue to fetch until it has exhausted the results from the API
```ruby
response = Vindicia::Repository::Account.all
```

### CODE REVIEW

File: lib/vindicia/repository/base.rb [hunk 1]
```diff
@@ -26,5 +46,16 @@ def self.route(id=nil)
       plural_class_name = ActiveSupport::Inflector.pluralize(class_name)
       [ '', plural_class_name, id ].compact.join('/')
     end
+
+    def self.cast(hash)
+      case hash['object']
+      #when 'Error'
```
Comment (jonstorer) [inline lib/vindicia/repository/base.rb]: I'll remove this before I merge, I want to start a discussion though... would you prefer that I return a response that has an error body? or throw the error?

they'd look like this:

```ruby
response = Vindicia::Repository::Account.first
if response.success?
  # do work with response
else
  error = response.body
  # do work with error
end
```
or
```ruby
begin
  response = Vindicia::Repository::Account.first
rescue Vindicia::Exception => error
  # do work with error
end
# do work with response
```

File: lib/vindicia/repository/base.rb [hunk 2]
```diff
@@ -1,9 +1,27 @@
...
+require 'addressable/uri'
 
 module Vindicia::Repository
   class Base
+    DEFAULT_LIMIT = 100.freeze
+
     def self.where(query = {})
-      Vindicia::Response::Collection.new(Vindicia::Request.new(:get, route, { query: { limit: 100 }.merge(query)}))
+      max = query.delete(:limit)
+      limit = [ ( max || DEFAULT_LIMIT ), DEFAULT_LIMIT ].min
```
Comment (kylessnell) [inline lib/vindicia/repository/base.rb]: why prevent a user from setting max to > 100?

Comment (jonstorer) [inline lib/vindicia/repository/base.rb]: Vindicia's API limits requests to 100 records at a time. 

I'm seeing the per request limit, not the total limit on the request.

File: lib/vindicia/repository/base.rb [hunk 3]
```diff
@@ -1,9 +1,27 @@
...
+      query = { limit: limit }.merge(query)
+
+      objects = []
+
+      begin
+        response = Vindicia::Request.new(:get, route, { query: query }).response
+        objects.concat(self.cast(response))
+      end while begin
+        if (response['next'] && (!max || objects.count < max))
```
Comment (kylessnell) [inline lib/vindicia/repository/base.rb]: `max.nil?`

File: lib/vindicia/repository/base.rb [hunk 4]
```diff
@@ -1,9 +1,27 @@
...
+      max = query.delete(:limit)
+      limit = [ ( max || DEFAULT_LIMIT ), DEFAULT_LIMIT ].min
+      query = { limit: limit }.merge(query)
+
+      objects = []
+
+      begin
+        response = Vindicia::Request.new(:get, route, { query: query }).response
+        objects.concat(self.cast(response))
+      end while begin
```
Comment (kylessnell) [inline lib/vindicia/repository/base.rb]: this is kinda gross

File: lib/vindicia/repository/base.rb (hunk 3 above)
Comment (jonstorer) [inline lib/vindicia/repository/base.rb]: Can do

File: lib/vindicia/repository/base.rb (hunk 4 above)
Comment (jonstorer) [inline lib/vindicia/repository/base.rb]: yeah, I know,,, help me make it better?

File: lib/vindicia/repository/base.rb (hunk 1 above)
Comment (kylessnell) [inline lib/vindicia/repository/base.rb]: 2nd one. let the user choose how to handle the error.

File: lib/vindicia/request.rb [hunk 5]
```diff
@@ -3,14 +3,15 @@ class Request
     include HTTParty
     format :json
     base_uri 'https://api.vindicia.com'
+    #debug_output $stdout
```
Comment (kylessnell) [inline lib/vindicia/request.rb]: 💀

File: spec/integration/product_spec.rb [hunk 6]
```diff
@@ -1,7 +1,6 @@
 require 'spec_helper'
 
-describe 'Product' do
-  before { Vindicia.test!  }
+describe 'Product' do before { Vindicia.test!  }
```
Comment (kylessnell) [inline spec/integration/product_spec.rb]: move before block to its own line. extra ws after !

File: spec/support/fixtures/products_two.json [hunk 7]
```diff
@@ -0,0 +1,75 @@
+{
+  "object": "List",
+  "data": [
+    {
+      "object": "Product",
+      "id": "2",
+      "vid": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
```
Comment (kylessnell) [inline spec/support/fixtures/products_two.json]: shouldn't these all have different vids

File: spec/unit/request_spec.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
+require 'spec_helper'
+
+describe Vindicia::Request do
+  context 'includes' do
+    subject { Vindicia::Request.new(:get, '/path') }
+
+    it 'HTTParty' do
+      expect(subject).to be_a HTTParty
```
Comment (kylessnell) [inline spec/unit/request_spec.rb]: 🎉

File: lib/vindicia/repository/base.rb (hunk 1 above)
Comment (jonstorer) [inline lib/vindicia/repository/base.rb]: 👍

File: spec/support/fixtures/products_two.json (hunk 7 above)
Comment (jonstorer) [inline spec/support/fixtures/products_two.json]: well, IRL yes,,, but for tests it doesn't happen. I'm not really enforcing that the api results are unique

**Pull request closed without merging**